contracts/hydro/src/contract.rs

```diff
@@ -715,71 +736,89 @@ fn vote(
-                (round_id, tranche_id, vote.prop_id),
-                &proposal,
-            )?;
-
-            // Add proposal's new power in PROPS_BY_SCORE
-            if proposal.power > Uint128::zero() {
-                PROPS_BY_SCORE.save(
+        )?;
+        match vote {
+            Some(vote) => {
+                // Load the proposal in the vote
+                let mut proposal =
+                    PROPOSAL_MAP.load(deps.storage, (round_id, tranche_id, vote.prop_id))?;
+
+                // Remove proposal's old power in PROPS_BY_SCORE
+                PROPS_BY_SCORE.remove(
                     deps.storage,
                     (
                         (round_id, proposal.tranche_id),
                         proposal.power.into(),
                         vote.prop_id,
                     ),
-                    &vote.prop_id,
+                );
+
+                remove_validator_shares_from_proposal(
+                    deps.storage,
+                    round_id,
+                    vote.prop_id,
+                    vote.time_weighted_shares.0,
+                    vote.time_weighted_shares.1,
                 )?;
-            }
 
-            // Delete vote
-            VOTE_MAP.remove(
-                deps.storage,
-                ((round_id, tranche_id), info.sender.clone(), lock_id),
-            );
+                // save the new power into the proposal
+                let total_power =
+                    get_total_power_for_proposal(deps.as_ref().storage, vote.prop_id)?;
+                proposal.power = total_power.to_uint_ceil(); // TODO: decide whether we need to round or represent as decimals
 
-            response = response.add_attribute(
-                format!("lock_id_{}_old_proposal_id", lock_id),
-                vote.prop_id.to_string(),
-            );
-        }
+                // Save the proposal
+                PROPOSAL_MAP.save(
+                    deps.storage,
+                    (round_id, tranche_id, vote.prop_id),
+                    &proposal,
+                )?;
+
+                // Add proposal's new power in PROPS_BY_SCORE
+                if proposal.power > Uint128::zero() {
+                    PROPS_BY_SCORE.save(
+                        deps.storage,
+                        (
+                            (round_id, proposal.tranche_id),
+                            proposal.power.into(),
+                            vote.prop_id,
+                        ),
+                        &vote.prop_id,
+                    )?;
+                }
+
+                // Delete vote
+                VOTE_MAP.remove(
+                    deps.storage,
+                    ((round_id, tranche_id), info.sender.clone(), lock_id),
+                );
+
+                response = response.add_attribute(
+                    format!("lock_id_{}_old_proposal_id", lock_id),
+                    vote.prop_id.to_string(),
+                );
+            }
+            None => {
```

Comment:
that's a smart way to do it, nice.

could you add a comment to the none branch, maybe, that this means that users can change their vote during a round, because we don't check this if users already voted?

Reply:
Yeah, it got me thinking at first, but Bernd did it in a very smart way. :) 
Will add the comment.